File: java/com/countrygamer/angel/client/gui/HUDOverlay.java
package com.countrygamer.angel.client.gui;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.client.event.RenderGameOverlayEvent;
import net.minecraftforge.client.event.RenderGameOverlayEvent.ElementType;

import org.lwjgl.opengl.GL11;

import com.countrygamer.angel.common.WeepingAngels;
import com.countrygamer.angel.common.extended.ExtendedAngelPlayer;
import com.countrygamer.core.Base.Plugin.ExtendedEntity;

import cpw.mods.fml.common.eventhandler.EventPriority;
import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

@SideOnly(Side.CLIENT)
public class HUDOverlay extends Gui {
	
	private Minecraft			mc;
	private static final int	BUFF_ICON_SIZE			= 9;
	public static final int		maxHUDHealth			= 20;
	
	public HUDOverlay(Minecraft mc) {
		super();
		this.mc = mc;
	}
	
	private static final ResourceLocation	health	= new ResourceLocation(WeepingAngels.PLUGIN_ID,
															"textures/gui/angelHealth.png");
	
	@SubscribeEvent(priority = EventPriority.NORMAL)
	public void renderGameOverlay(RenderGameOverlayEvent.Post event) {
		if (event.isCancelable() || event.type != ElementType.EXPERIENCE) {
			return;
		}
		
		int width = event.resolution.getScaledWidth();
		int xPos = (int) (width / 2 * 0.04) + width / 2;
		int height = event.resolution.getScaledHeight();
		int yPos = (height / 2 + height / 4) + (int) (height / 4 * 0.19);
		ScaledResolution scaledresolution = new ScaledResolution(this.mc.gameSettings,
				this.mc.displayWidth, this.mc.displayHeight);
		int k = scaledresolution.getScaledWidth();
		int l = scaledresolution.getScaledHeight();
		
		ExtendedAngelPlayer angelPlayer = (ExtendedAngelPlayer) ExtendedEntity.getExtended(
				this.mc.thePlayer, ExtendedAngelPlayer.class);
		float angelHealth = angelPlayer.getAngelHealth();
		if (angelHealth > 0.0F) {
			
			GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
			GL11.glDisable(GL11.GL_LIGHTING);
			this.mc.getTextureManager().bindTexture(health);
			
			for (int i = 0; i < HUDOverlay.maxHUDHealth / 2; i++) {
				this.drawTexturedModalRect(xPos + (i * 9), yPos, 0, 9, BUFF_ICON_SIZE,
						BUFF_ICON_SIZE);
			}
			
			int fullHearts = (int) Math.floor(angelHealth / 2);
			int halfHearts = (int) (angelHealth - (fullHearts * 2));
			int xPosLastFullHeart = 0;
			for (int i = 0; i < fullHearts; i++) {
				xPosLastFullHeart = xPos + (i * 9);
				this.drawTexturedModalRect(xPosLastFullHeart, yPos, 0, 0, BUFF_ICON_SIZE,
						BUFF_ICON_SIZE);
			}
			for (int i = 0; i < halfHearts; i++) {
				if (xPosLastFullHeart <= 0)
					this.drawTexturedModalRect(xPos + (i * 9), yPos, 9, 0, BUFF_ICON_SIZE,
							BUFF_ICON_SIZE);
				else
					this.drawTexturedModalRect(xPosLastFullHeart + ((i + 1) * 9), yPos, 9, 0,
							BUFF_ICON_SIZE, BUFF_ICON_SIZE);
			}
			
			this.renderBlur(k, l, angelHealth / (float) (HUDOverlay.maxHUDHealth));
		}
	}
	
	protected static final ResourceLocation	blackBlur	= new ResourceLocation(
																WeepingAngels.PLUGIN_ID,
																"textures/gui/blackBlur.png");
	
	protected void renderBlur(int x, int y, float alpha) {
		
		GL11.glEnable(GL11.GL_BLEND);
		GL11.glDisable(GL11.GL_DEPTH_TEST);
		GL11.glDepthMask(false);
		GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
		GL11.glColor4f(1.0F, 1.0F, 1.0F, alpha);
		GL11.glDisable(GL11.GL_ALPHA_TEST);
		
		this.mc.getTextureManager().bindTexture(blackBlur);
		Tessellator tessellator = Tessellator.instance;
		tessellator.startDrawingQuads();
		tessellator.addVertexWithUV(0.0D, (double) y, -90.0D, 0.0D, 1.0D);
		tessellator.addVertexWithUV((double) x, (double) y, -90.0D, 1.0D, 1.0D);
		tessellator.addVertexWithUV((double) x, 0.0D, -90.0D, 1.0D, 0.0D);
		tessellator.addVertexWithUV(0.0D, 0.0D, -90.0D, 0.0D, 0.0D);
		tessellator.draw();
		
		GL11.glDisable(GL11.GL_BLEND);
		
	}
	
}

File: src/main/scala/com/temportalist/weepingangels/client/gui/HUDOverlay.scala
package com.temportalist.weepingangels.client.gui

import com.temportalist.origin.library.client.utility.{Rendering, TessRenderer}
import com.temportalist.origin.wrapper.common.extended.ExtendedEntityHandler
import com.temportalist.weepingangels.common.extended.AngelPlayer
import com.temportalist.weepingangels.common.{WAOptions, WeepingAngels}
import cpw.mods.fml.common.eventhandler.{EventPriority, SubscribeEvent}
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.{Gui, ScaledResolution}
import net.minecraft.util.ResourceLocation
import net.minecraftforge.client.event.RenderGameOverlayEvent
import net.minecraftforge.client.event.RenderGameOverlayEvent.ElementType
import org.lwjgl.opengl.GL11

/**
 *
 *
 * @author TheTemportalist
 */
@SideOnly(Side.CLIENT)
object HUDOverlay extends Gui() {

	val mc: Minecraft = Minecraft.getMinecraft
	val iconSize: Int = 9

	val healthTexture: ResourceLocation = new ResourceLocation(WeepingAngels.MODID,
		"textures/gui/angelHealth.png")
	val blackout: ResourceLocation = new ResourceLocation(WeepingAngels.MODID,
		"textures/gui/blackBlur.png"
	)

	var resolution: ScaledResolution = null

	@SubscribeEvent(priority = EventPriority.NORMAL)
	def renderOverlay(event: RenderGameOverlayEvent.Post): Unit = {

		if (event.isCanceled) {
			//return
		}

		val angelPlayer: AngelPlayer = ExtendedEntityHandler
				.getExtended(this.mc.thePlayer, classOf[AngelPlayer]).asInstanceOf[AngelPlayer]

		if (!angelPlayer.converting()) return

		val angelHealth: Float = angelPlayer.getAngelHealth()
		//System.out.println(angelHealth)

		val width: Int = event.resolution.getScaledWidth
		val height: Int = event.resolution.getScaledHeight

		// Used for transparency because it is immediately before crosshairs, which precedes all hud things
		if (event.`type` == ElementType.HELMET) {
			this.renderBlackoutWithAlpha(angelPlayer.getOpacityForBlackout(), width, height)
		}

		// Triggered to render the angel health

		if (event.`type` == ElementType.FOOD) {
			GL11.glPushMatrix()
			GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F)
			GL11.glDisable(GL11.GL_LIGHTING)
			GL11.glEnable(GL11.GL_BLEND)

			val left: Int = width / 2 - 91 + 100
			val top: Int = height - 39 - 10

			val healthBarX: Int = left + 0
			val healthBarY: Int = top

			Rendering.bindResource(this.healthTexture)

			// Draw background hearts
			for (i <- 0 until WAOptions.maxAngelHealth / 2) {
				this.drawTexturedModalRect(healthBarX + (i * 8), healthBarY, 0, 9, this.iconSize,
					this.iconSize)
			}

			val fullHearts: Int = Math.floor(angelHealth / 2.0F).asInstanceOf[Int]
			val halfHearts: Int = Math.floor(angelHealth - (fullHearts * 2)).asInstanceOf[Int]

			for (i <- 0 until fullHearts) {
				this.drawTexturedModalRect(healthBarX + (i * 8), healthBarY, 0, 0, this.iconSize,
					this.iconSize)
			}

			for (i <- 0 until halfHearts) {
				this.drawTexturedModalRect(healthBarX + (fullHearts * 8) + (i * 8), healthBarY, 9,
					0, this.iconSize, this.iconSize)
			}

			GL11.glPopMatrix()
		}

	}

	def renderBlackoutWithAlpha(alpha: Float, width: Int, height: Int): Unit = {
		this.renderBlackoutWithAlpha(alpha, 0.0D, 0.0D, width, height)
	}

	def renderBlackoutWithAlpha(alpha: Float, x: Double, y: Double, width: Double,
			height: Double): Unit = {
		GL11.glPushMatrix()

		GL11.glEnable(GL11.GL_BLEND)
		GL11.glDisable(GL11.GL_DEPTH_TEST)
		GL11.glDepthMask(false)
		GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA)

		GL11.glColor4f(1.0F, 1.0F, 1.0F, alpha)

		GL11.glDisable(GL11.GL_ALPHA_TEST)

		Rendering.bindResource(this.blackout)

		TessRenderer.startQuads()
		TessRenderer.addVertex(x + 0.0D, y + height, -90.0D, 0.0D, 1.0D)
		TessRenderer.addVertex(x + width, y + height, -90.0D, 1.0D, 1.0D)
		TessRenderer.addVertex(x + width, y + 0.0D, -90.0D, 1.0D, 0.0D)
		TessRenderer.addVertex(x + 0.0D, y + 0.0D, -90.0D, 0.0D, 0.0D)
		TessRenderer.draw()

		GL11.glDepthMask(true)
		GL11.glEnable(GL11.GL_DEPTH_TEST)
		GL11.glDisable(GL11.GL_BLEND)

		GL11.glPopMatrix()
	}

}
